### PR TITLE
PUBDEV-8180: uniform Python API to retrieve trained model(s)/leader

### PR DESCRIPTION
**Pull request by sebhrusen**
https://h2oai.atlassian.net/browse/PUBDEV-8180

### CODE REVIEW

File: h2o-py/h2o/grid/grid_search.py [hunk 1]
```diff
@@ -77,9 +42,44 @@ class H2OGridSearch(h2o_meta(Keyed)):
...
+                ...             "stopping_tolerance": 0.00001}
+        :param export_checkpoints_dir: Directory to automatically export the grid and its models to.
+        :param recovery_dir: When specified, the grid and all necessary data (frames, models) will be saved to this
+            directory (use HDFS or other distributed file-system). Should the cluster crash during training, the grid
+            can be reloaded from this directory via ``h2o.load_grid``, and training can be resumed.
+        :param parallelism: Level of parallelism during grid model building. 1 = sequential building (default). 
+             Use the value of 0 for adaptive parallelism - decided by H2O. Any number > 1 sets the exact number of models
+             built in parallel.
+        :returns: a new H2OGridSearch instance
+        """
```
Comment (sebhrusen) [inline h2o-py/h2o/grid/grid_search.py:82]: just moved this doc where it belongs

File: h2o-py/h2o/utils/metaclass.py [hunk 2]
```diff
@@ -201,6 +200,7 @@ def __call__(self, clz):
         clz._bc = self
         new_clz = None
 
+        @wraps(clz.__init__)
```
Comment (sebhrusen) [inline h2o-py/h2o/utils/metaclass.py:203]: small fix here to ensure that the init signature and doc appears correctly when printing help in ipython/jupyter: noticed this when fixing the GridSearch docstring.

**Comment by michalkurka** [comment]
@sebhrusen looks very good, I have a question about GridSearch vs AutoML

GridSearch returns all models while AutoML returns the leader

how do we justify that? the concept is not dissimilar?

**Comment by sebhrusen** [comment]
@michalkurka I'm glad you ask, because I was wondering for a few minutes what to do with grids.

The main goal of AutoML is to produce one good model, so I think we all agree that it makes returning the leader there.

We could consider that the goal of grids is also to produce one good model at the end, but currently grids doesn't provide any API to select the best model among all, so it appears that our implementation de facto doesn't aim at producing one single model…
We could add this API to grids, but it would be out-of-scope for this ticket, so if we plan to add a model selection API to grids, I would suggest that we keep `train` method for grids as before (not returning anything), and we will make it return a single model only then.

wdyt?

**Comment by sebhrusen** [comment]
@michalkurka second thought: `R` return a grid  instance,  so maybe we should just return `self` in case of grids as well.

**Comment by ledell** [comment]
@sebhrusen 
> We could consider that the goal of grids is also to produce one good model at the end, but currently grids don't provide any API to select the best model among all, so it appears that our implementation de facto doesn't aim at producing one single model… 

I like the idea that you could train and get the top model from a grid in one-line of code, just like AutoML.  I think it's a good idea!  It seems like grid already comes pre-sorted, using a default metric.  In binary classification, its seems to be sorted by logloss.  Usually I use  `.get_grid(sort_by="auc")` but it seems that we could re-use the single-model leader concept here, and then we could use grids in pipelines.

Example (current functionality):

```
import h2o
from h2o.estimators.gbm import H2OGradientBoostingEstimator
from h2o.grid.grid_search import H2OGridSearch

h2o.init()

train = h2o.import_file("https://s3.amazonaws.com/erin-data/higgs/higgs_train_10k.csv")
test = h2o.import_file("https://s3.amazonaws.com/erin-data/higgs/higgs_test_5k.csv")
x = train.columns
y = "response"
x.remove(y)
train[y] = train[y].asfactor()
test[y] = test[y].asfactor()



# Specify GBM hyperparameters for the grid
hyper_params = {"learn_rate": [0.01, 0.03],
                "max_depth": [3, 4, 5, 6, 9],
                "sample_rate": [0.7, 0.8, 0.9, 1.0],
                "col_sample_rate": [0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8]}
search_criteria = {"strategy": "RandomDiscrete", "max_models": 3, "seed": 1}

# Train the grid
grid = H2OGridSearch(model=H2OGradientBoostingEstimator(ntrees=10,
                                                        seed=1,
                                                        nfolds=5),
                     hyper_params=hyper_params,
                     search_criteria=search_criteria,
                     grid_id="gbm_grid_binomial")
grid.train(x=x, y=y, training_frame=train)

grid_auc = grid.get_grid(sort_by='auc', decreasing=True)
```


Currently sorted by logloss by default:
```
In [26]: grid
Out[26]:     col_sample_rate learn_rate max_depth sample_rate                  model_ids             logloss
0               0.4       0.03         3         0.7  gbm_grid_binomial_model_3  0.6670120129698885
1               0.2       0.03         4         0.8  gbm_grid_binomial_model_2  0.6746770622641063
2               0.7       0.01         5         0.9  gbm_grid_binomial_model_1  0.6769884389744855

In [28]: grid_auc
Out[28]:     col_sample_rate learn_rate max_depth sample_rate                  model_ids                 auc
0               0.7       0.01         5         0.9  gbm_grid_binomial_model_1  0.7462602720068559
1               0.4       0.03         3         0.7  gbm_grid_binomial_model_3  0.7335792493670467
2               0.2       0.03         4         0.8  gbm_grid_binomial_model_2  0.7140694557755547
```

**Comment by sebhrusen** [comment]
@ledell @michalkurka I look at the backend and I can confirm what Erin says, the models are automatically sorted (for some reason, this logic is done in `GridSchemaV99`, which is really weird…).
And here are the default sorting metrics:
```
  public GridSortBy getDefaultGridSortBy() {
    if (! isSupervised())
      return null;
    else if (_output.nclasses() > 1)
      return GridSortBy.LOGLOSS;
    else
      return GridSortBy.RESDEV;
  }
 ```
Also, the `sort_by` param suggested could easily be added to the constructor: the logic is already there in the backend, it's the `sort_by` param is just not currently used when fetching the grid on the job completion… I won't  change this in this PR though.

Which leaves us with 2 possibilities as `train` result:
1. return the grid itself: justification being that it has a predict method, and metrics methods similar  to the ones provided with a single model, the difference being that those methods predict/compute metrics on all models of the grid.
2. return the first model: justification is to be consistent with AutoML, and estimator APIs.  The first model will always use the default sorting for now (could be improved later if we add the `sort_by` param to the grid constructor). The only glitch here is for unsupervised models: we can't justify picking the first model.

Time to vote, 1 or 2?

**Comment by sebhrusen** [comment]
@michalkurka @ledell are we ok with this one? there's nothing blocking the PR otherwise, last build was Python-green, I just had to trigger a new build for the PR metadata check (now that this is fixed for client-only PRs).

**Review by michalkurka — APPROVED** [review]
LGTM - nice change

**Review by koniecsveta — APPROVED** [review]

**Pull request merged**